test/piv_test.go

```diff
@@ -0,0 +1,129 @@
+	ctx := context.Background()
+
+	Confirm = func(_ string) bool { return true }
+	must(ResetKeyCmd(ctx), t)
+
+	// The key should be the default management key so this should fail
+	mustErr(SetManagementKeyCmd(ctx, "foobar", "newkey", false), t)
+	must(SetManagementKeyCmd(ctx, "", "newkey", false), t)
+
+	// Now it should fail with the wrong key
+	mustErr(SetManagementKeyCmd(ctx, "", "otherkey", false), t)
+	// But pass if we use the right key
+	must(SetManagementKeyCmd(ctx, "newkey", "otherkey", false), t)
+
+	// Reset and try a random key too!
+	must(ResetKeyCmd(ctx), t)
+	must(SetManagementKeyCmd(ctx, "", "", true), t)
+	mustErr(SetManagementKeyCmd(ctx, "", "", true), t)
+}
+
+func TestSetPUKCmd(t *testing.T) {
+	ctx := context.Background()
+
+	Confirm = func(_ string) bool { return true }
+	must(ResetKeyCmd(ctx), t)
+
+	// The PUK should be the default key so this should fail
+	mustErr(SetPukCmd(ctx, "11111111", "12121212"), t)
+	must(SetPukCmd(ctx, "", "12121212"), t)
+
+	// Now it should fail with the wrong key
+	mustErr(SetPukCmd(ctx, "", "43214321"), t)
+	// But pass if we use the right key
+	must(SetPukCmd(ctx, "12121212", "43214321"), t)
+}
+
+func TestSetPinCmd(t *testing.T) {
+	ctx := context.Background()
+
+	Confirm = func(_ string) bool { return true }
+	must(ResetKeyCmd(ctx), t)
+
+	// The PIN should be the default PIN so this should fail
+	mustErr(SetPinCmd(ctx, "111111", "222222"), t)
+	must(SetPinCmd(ctx, "", "222222"), t)
+
+	// Now it should fail with the wrong key
+	mustErr(SetPinCmd(ctx, "333333", "444444"), t)
+	// But pass if we use the right key
+	must(SetPinCmd(ctx, "222222", "111111"), t)
+}
+
+func TestUnblockCmd(t *testing.T) {
+	ctx := context.Background()
+
+	Confirm = func(_ string) bool { return true }
+	must(ResetKeyCmd(ctx), t)
+
+	// Set a PUK
+	must(SetPukCmd(ctx, "", "43214321"), t)
+	// Set the pin to something, then lock the device by trying the wrong one too many times.
+	must(SetPinCmd(ctx, "", "111111"), t)
+
+	for i := 0; i < 5; i++ {
+		mustErr(SetPinCmd(ctx, "222222", "333333"), t)
+	}
+
+	// Now even with the right PIN it should be stuck
+	mustErr(SetPinCmd(ctx, "111111", "222222"), t)
+
+	// But we can unblock it
+	must(UnblockCmd(ctx, "43214321", "222222"), t)
+	must(SetPinCmd(ctx, "222222", "333333"), t)
+}
+
+func TestGenerateKeyCmd(t *testing.T) {
+	ctx := context.Background()
+
+	Confirm = func(_ string) bool { return true }
+	must(ResetKeyCmd(ctx), t)
+
+	// This should work with the default key
+	must(GenerateKeyCmd(ctx, "", false), t)
+
+	// Set the key to something other than the default
+	must(SetManagementKeyCmd(ctx, "", "mynewkey", false), t)
+	// Now this should fail
+	mustErr(GenerateKeyCmd(ctx, "", false), t)
+	// Unless we use the right key
+	must(GenerateKeyCmd(ctx, "mynewkey", false), t)
+
+	// Now if we use a random key it should set a new one
+	must(GenerateKeyCmd(ctx, "mynewkey", true), t)
+	// The old one shouldn't work.
+	mustErr(GenerateKeyCmd(ctx, "mynewkey", false), t)
+}
```

Comment:
would it be useful to have a `TestAttestation` which basically tests the flow described in `TOKENS.md`?

Reply:
ooh good call. I didn't want to add a test without having any way to check it against something, but that's perfect!

Reply:
BOOM

Reply:
WOOT